# The Book Keeper's Apprentice - Dicts & List
 

In [1]:
# Importing libraries
import time
from pprint import pprint

# Importing data
from helper import books_dict, authors_list, books_list

In [2]:
# Investigate the `books_dict` dictionary and the two lists
# `authors_list`, # `books_list`
print(authors_list[:10])
print(books_list[:10])
len(authors_list)
len(books_list)

['Michael W. Dean', 'Patricia Schultz', 'Dick Francis', 'Isaac Asimov', 'Pablo Neruda', 'Ralph Ginzburg', 'Michael Ochs', 'Marthe Le Van', 'Owen    Jones', 'George Balanchine']
['$30 Film School: How to Write  Direct  Produce  Shoot  Edit  Distribute  Tour With  and Sell Your Own No-Budget Digital Movie', '1 000 Places to See Before You Die', '10 lb Penalty', '100 Great Fantasy Short Short Stories', '100 Love Sonnets', '100 Years of Lynchings', '1000 Record Covers', '1000 Rings: Inspiring Adornments for the Hand', '1001 Illuminated Initial Letters: 27 Full-Color Plates', '101 Stories of the Great Ballets: The Scene-by-Scene Stories of the Most Popular Ballets  Old and New']


10019

### ⏸ How many elements are there in the two lists provided?


#### A. 20019
#### B. 10220
#### C. 10019

HINT: Use `len()` on `authors_list` and `books_list`

In [3]:
### edTest(test_0) ###
## DO NOT REMOVE ABOVE LINE ##

# Submit an answer choice as a string below 
# (eg. if you choose option C, put 'C')
your_answer = 'C'

## Part 1 - Searching for Book 1: `"1 000 Places to See Before You Die"`

#### Using `books_list`, fetch the author name for `"1 000 Places to See Before You Die"` *(exact string match)*

In [4]:
### edTest(test_1) ###
## DO NOT REMOVE ABOVE LINE ##

# Let's start the clock
start_time = time.time()

print("Searching through the list...")

# Enter the book title
book_title = "1 000 Places to See Before You Die"

# Go through the books list and find the index associated with book_title
# Find the author name using that index on authors_list
for i in range(len(books_list)):
    if books_list[i] == book_title:
        author_name = authors_list[i]
        break

print(f"The author of the book is {author_name}")
dict_duration_book_1 = time.time() - start_time
print(f"--- Searching took {dict_duration_book_1:.5f} seconds --- ")

Searching through the list...
The author of the book is Patricia Schultz
--- Searching took 0.00141 seconds --- 


#### Using `books_dict`, fetch the author name for`"1 000 Places to See Before You Die"`

Search for the books from the dictionary and fetch the author name.

In [5]:
### edTest(test_2) ###
## DO NOT REMOVE ABOVE LINE ##

# Let's start the clock
start_time = time.time()

print("Searching through the dict...")

# Book title
book_title = "1 000 places to see before you die"

# Fetch the value from `books_dict` using the title key
author_name = authors_list[i]

print(f"The author of the book is {author_name}")

# Let's measure how much time it took
dict_duration_book_1 = time.time() - start_time
print(f"--- Searching took {dict_duration_book_1:.5f} seconds --- ")

Searching through the dict...
The author of the book is Patricia Schultz
--- Searching took 0.00021 seconds --- 


### Head to Head
Let's compare the amount of time it took for list search vs dictionary search

In [6]:
# Calculate the ratio of the time taken by list search to dictionary search
speed_ratio = ___
print(f"Ratio of list search to dictionary search is {speed_ratio:.3f}")

Please note that in this case, your `book_index` was relatively small. This means that the book is placed in earlier positions. (Check value of `i` in the list search loop above to verify).

## Part 2 - Searching for book 2: `"Youth in Revolt: The Journals of Nick Twisp"`

#### Using `books_list`, fetch the author name of "Youth in Revolt: The Journals of Nick Twisp"

In [6]:
### edTest(test_3) ###
## DO NOT REMOVE ABOVE LINE ##

# Let's start the clock
start_time = time.time()

print("Searching through the list...")

# Title of the book
book_title = "Youth in Revolt: The Journals of Nick Twisp"

# Go through the books list and find the index associated with book_title
# Find the author name using that index on authors_list
for i in range(len(books_list)):
    if books_list[i]==book_title:
        author_name=authors_list[i]
        break

# Print the name of the author of the book
print(f"The author of the book is {author_name}")

# Let's measure how much time it took
list_duration_book_2 = time.time() - start_time
print(f"--- Searching took {list_duration_book_2:.5f} seconds ---")

Searching through the list...
The author of the book is C.D. Payne
--- Searching took 0.00337 seconds ---


#### Using `books_dict`, fetch the author name of `"Youth in Revolt: The Journals of Nick Twisp"`

In [7]:
### edTest(test_4) ###
## DO NOT REMOVE ABOVE LINE ##

# Let's start the clock
start_time = time.time()
start_time = time.time()

print("Searching through the list...")

# Title of the book
book_title = "Youth in Revolt: The Journals of Nick Twisp"

# Go through the books list and find the index associated with book_title
# Find the author name using that index on authors_list
for i in range(len(books_list)):
    if books_list[i]==book_title:
        author_name=authors_list[i]
        break    # Print the name of the author of the book
print(f"The author of the book is {author_name}")

# Let's measure how much time it took
list_duration_book_2 = time.time() - start_time
print(f"--- Searching took {list_duration_book_2:.5f} seconds ---")


Searching through the list...
The author of the book is C.D. Payne
--- Searching took 0.00247 seconds ---


In [9]:
# Calculate the ratio of the time taken by list search to dictionary search
speed_ratio = ___
print(f"Ratio of list search to dictionary search is {speed_ratio:.3f}")

Do you observe the relation between position and the time it is taking to fetch values from that position? How is it differing for a list vs. a dictionary?

### ⏸ Which took lesser time - searching through a list or searching through the dictionary? Why?

#### A. Dictionary was much faster in the second search
#### B. List was much faster in the second search
#### C. Dictionary was much faster in the first search


In [8]:
### edTest(test_5) ###
## DO NOT REMOVE ABOVE LINE ##

# Submit an answer choice as a string below 
# (eg. if you choose option C, put 'C')
your_answer = 'A'